server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,596 @@
+        )
+        tarball_dirs = list()
+        unpacking_tarball_dirs = list()
+
+        if not Path(ihier.config.ARCHIVE, controller).is_dir():
+            # Skip incoming controller directories that don't have an $ARCHIVE
+            # directory, handled in another part of the audit.
+            continue
+
+        for dirent in direct_entries:
+            dir_path = Path(dirent)
+            if (
+                dir_path.is_dir()
+                and not dir_path.name.endswith(".unpack")
+                and not len(os.listdir(dir_path)) == 0
+            ):
+                tarball_dirs.append(dir_path.name)
+            elif (
+                dir_path.is_dir()
+                and not dir_path.name.endswith(".unpack")
+                and len(os.listdir(dir_path)) == 0
+            ):
+                ihier.add_unexpected_controllers(
+                    Hierarchy.EMPTY_TARBALL_DIRS, controller, dir_path.name
+                )
+            elif (
+                dir_path.is_dir()
+                and dir_path.name.endswith(".unpack")
+                and len(os.listdir(dir_path)) == 0
+            ):
+                unpacking_tarball_dirs.append(dir_path.name)
+            elif dir_path.is_symlink():
+                ihier.add_unexpected_controllers(
+                    Hierarchy.TARBALL_LINKS, controller, dir_path.name
+                )
+
+        if tarball_dirs:
+            verify_tar_dirs(
+                ihier, tarball_dirs, ihier.add_unexpected_controllers, controller
+            )
+
+        if unpacking_tarball_dirs:
+            verify_tar_dirs(
+                ihier,
+                unpacking_tarball_dirs,
+                ihier.add_unexpected_controllers,
+                controller,
+            )
+
+    return 0
+
+
+# Results Hierarchy
+def verify_user_arg(rhier, mdlogcfg, controller, path):
+    """fetch user from the config file and verifies its validity"""
+
+    if rhier.path.parent == rhier.config.USERS:
+        """We are reviewing a user tree, so check the user in
+        the configuration.  Version 002 agents use the
+        metadata log to store a user as well.
+        """
+        user_arg = rhier.path.name
+        user = ""
+        try:
+            user = mdlogcfg.get("run", "user")
+        except NoSectionError:
+            pass
+        except NoOptionError:
+            pass
+        if not user:
+            """No user in the metadata.log of the tar ball, but
+            we are examining a link in the user tree that
+            does not have a configured user, report it.
+            """
+            rhier.add_unexpected_controllers(
+                Hierarchy.UNEXPECTED_USER_LINKS, controller, path
+            )
+        elif user_arg != user:
+            """Configured user does not match the user tree in
+            which we found the link."""
+            rhier.add_unexpected_controllers(
+                Hierarchy.WRONG_USER_LINKS, controller, path
+            )
```

Comment:
You could use direct_entries.extend(files) instead of iterating through each list (and similarly for the dirs list).

Also, you seem to have taken a shortcut on this method in other places by just globbing a base_dir/* ... not sure why you're doing this two different ways?